### PR TITLE
WIP: HEARTBEAT and commander failsafe handling cleanup

### PR DESCRIPTION
**Pull request by dagar**
For discussion of expanded failsafe handling WRT mavlink HEARTBEAT (https://github.com/PX4/Firmware/pull/10841).

 - mavlink update telemetry_status with all HEARTBEAT timestamps
 - let commander decide which HEARTBEATs to use
 - optimizations to avoid unnecessary state and subscribing to all telemetry_status and iridiumsbd instances (saves ~1.5kB of RAM)
 - simplifies convoluted commander data link handling

 Related
 - https://github.com/PX4/Firmware/pull/10841
 - https://github.com/PX4/Firmware/pull/10194
 - https://github.com/PX4/Firmware/pull/9473
 - https://github.com/PX4/Firmware/pull/7985

### CODE REVIEW

File: src/modules/commander/Commander.hpp [hunk 1]
```diff
@@ -98,6 +98,17 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
 private:
 
 	DEFINE_PARAMETERS(
+
+		(ParamInt<px4::params::NAV_DLL_ACT>) _datalink_loss_action,
```
Comment (TSC21) [inline src/modules/commander/Commander.hpp]: If this came to the commander param list, why not change the suffix to `COM`?

File: src/modules/commander/Commander.hpp [hunk 2]
```diff
@@ -98,6 +98,17 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
...
+		(ParamInt<px4::params::COM_DL_LOSS_T>) _datalink_loss_threshold,
+		(ParamInt<px4::params::COM_DL_REG_T>) _datalink_regain_threshold,
+
+		(ParamInt<px4::params::COM_HLDL_LOSS_T>) _high_latency_datalink_loss_threshold,
+		(ParamInt<px4::params::COM_HLDL_REG_T>) _high_latency_datalink_regain_threshold,
+
+		(ParamInt<px4::params::NAV_RCL_ACT>) _rc_loss_action,
```
Comment (TSC21) [inline src/modules/commander/Commander.hpp]: Same thing here

File: src/modules/commander/Commander.hpp [hunk 3]
```diff
@@ -98,6 +98,17 @@ class Commander : public ModuleBase<Commander>, public ModuleParams
 private:
 
 	DEFINE_PARAMETERS(
+
+		(ParamInt<px4::params::NAV_DLL_ACT>) _datalink_loss_action,
+		(ParamInt<px4::params::COM_DL_LOSS_T>) _datalink_loss_threshold,
+		(ParamInt<px4::params::COM_DL_REG_T>) _datalink_regain_threshold,
+
+		(ParamInt<px4::params::COM_HLDL_LOSS_T>) _high_latency_datalink_loss_threshold,
```
Comment (TSC21) [inline src/modules/commander/Commander.hpp]: so for an `_obc_datalink_loss`, we need to add new params right?

File: src/modules/commander/Commander.cpp [hunk 4]
```diff
@@ -3760,202 +3744,107 @@ bool Commander::preflight_check(bool report)
...
+					if (status.data_link_lost) {
+						if (hrt_elapsed_time(&_datalink_lost) > (_datalink_regain_threshold.get() * 1_s)) {
+							status.data_link_lost = false;
+							status_changed = true;
+						}
+					}
+					break;
+				case (telemetry_status_s::MAV_TYPE_ONBOARD_CONTROLLER):
+					_datalink_last_heartbeat_onboard_controller = telemetry.heartbeat_time;
+					// TODO: FYI @baumanta
```
Comment (TSC21) [inline src/modules/commander/Commander.cpp]: @baumanta

File: src/modules/commander/Commander.hpp (hunk 1 above)
Comment (dagar) [reply src/modules/commander/Commander.hpp]: They've already effectively been commander params for several releases. The only reason I didn't do the rename is that users will lost existing settings. This is the type of thing we can do for a release with proper migration documentation.

File: src/modules/commander/Commander.hpp (hunk 2 above)
Comment (dagar) [reply src/modules/commander/Commander.hpp]: See previous comment.

File: src/modules/commander/Commander.hpp (hunk 3 above)
Comment (dagar) [reply src/modules/commander/Commander.hpp]: One more time for clarity, how would you like it to work?

Comment (TSC21) [reply src/modules/commander/Commander.hpp]: For example: I have a GCS link and a OBC link. I want the OBC link, if failing, to trigger a failsafe, but if the GCS link fails, I just want it to send a mavlink_log warning me, but not triggering a failsafe. Or, I may want them to both trigger a failsafe, or want that none trigger a failsafe but at least warn me this also applies to the high latency case.

Imagine we have 3 links. I would like to state which ones do trigger failsafes in the case of failure, but want to all of them to tell at least warn me in the case of a failsafe.

Comment (TSC21) [reply src/modules/commander/Commander.hpp]: It can make sense to be more atomic and say: OBC triggers a RTL failsafe, while GCS triggers a land failsafe

Comment (dagar) [reply src/modules/commander/Commander.hpp]: TODO: we need to be careful for the scenario where a single mavlink connection will have multiple HEARTBEATs (eg companion also routing to GCS). I'll make sure that each received HEARTBEAT results in a telemetry_status publication.

**Comment by TSC21** [comment]
@baumanta your commit looks really good. But it's failing for a specific target: http://ci.px4.io:8080/blue/organizations/jenkins/PX4_misc%2FFirmware-compile/detail/pr-heartbeat_failsafes/5/pipeline#step-533-log-407

**Comment by MaEtUgR** [comment]
@dagar I quickly checked, so far no conflict with #10907 😅

**Comment by baumanta** [comment]
@dagar are there any updates on this?

**Comment by dagar** [comment]
TODO: double check the iridium behavior

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 5]
```diff
@@ -1949,7 +1949,7 @@ MavlinkReceiver::handle_message_heartbeat(mavlink_message_t *msg)
 		mavlink_msg_heartbeat_decode(msg, &hb);
 
 		/* ignore own heartbeats, accept only heartbeats from GCS */
-		if (msg->sysid != mavlink_system.sysid) {
+		if (msg->sysid != mavlink_system.sysid || hb.type == MAV_TYPE_ONBOARD_CONTROLLER) {
```
[the remark this answers is not in the document]
Comment (dagar) [reply src/modules/mavlink/mavlink_receiver.cpp]:
```suggestion

```

**Comment by dagar** [comment]
I'll try to resurrect this and get it in incrementally. 
Part 1 - https://github.com/PX4/Firmware/pull/11454

**Comment by jkflying** [comment]
I guess this is all in now and can be closed?

**Comment by dagar** [comment]
Yes, thanks everyone!

**Pull request closed without merging**